src/Streamly/Internal/Data/Stream/StreamD/Exception.hs

```diff
@@ -349,3 +352,53 @@ _handle f (Stream step state) = Stream step' (Left state)
             Yield x s -> return $ Yield x (Right (Stream step1 s))
             Skip s    -> return $ Skip (Right (Stream step1 s))
             Stop      -> return Stop
+
+data RetryState m s1 s2 = WithOldStream m s1 | WithNewStream s2
+
+-- | See 'Streamly.Internal.Data.Stream.IsStream.retry'
+--
+{-# INLINE_NORMAL retry #-}
+retry
+    :: forall e m a. (Exception e, Ord e, MonadCatch m)
+    => Map e Int
+       -- ^ map from exception to retry count
+    -> (e -> Stream m a)
+       -- ^ default handler for those exceptions that are not in the map
+    -> Stream m a
+    -> Stream m a
+retry emap0 han0 (Stream step0 state0) = Stream step state
+
+    where
+
+    state = WithOldStream emap0 state0
+
+    {-# INLINE_LATE step #-}
+    step gst (WithOldStream emap st) = do
+        -- XXX gtry? Whats the difference?
```

Comment:
I don't understand this.

Reply:
Me neither. I think it was essentially about `ExceptionMonad vs MonadCatch`